src/entity.ts

```diff
@@ -0,0 +1,53 @@
+import { Observable } from 'rxjs'
+import { first } from 'rxjs/operators'
+import {
+  Arc,
+  IApolloQueryOptions,
+  Address
+} from './index'
+
+interface IEntityState {
+  id: string
+}
+
+export interface IEntityRef<Entity> {
+  id: Address,
+  entity: Entity
+}
+
+export abstract class Entity<TEntityState extends IEntityState> {
+  id: string
+  context: Arc
+  coreState: TEntityState | undefined
+
+  constructor(
+    context: Arc,
+    idOrOpts: string|TEntityState,
+  ) {
+    this.context = context
+    if (typeof idOrOpts === 'string') {
+      this.id = idOrOpts.toLowerCase()
+    } else {
+      this.id = idOrOpts.id
+      this.setState(idOrOpts)
+    }
+  }
+
+  abstract state(apolloQueryOptions: IApolloQueryOptions): Observable<TEntityState>
+
+  setState(entityState: TEntityState): void {
+    this.coreState = entityState
+  }
+
+  async fetchState(apolloQueryOptions: IApolloQueryOptions = {}, refetch?: boolean): Promise<TEntityState> {
```

Comment:
let's be explicitly about access modifiers for functions, this should have `public` in-front of it